Include/methodobject.h

```diff
@@ -13,7 +13,7 @@ extern "C" {
 
 PyAPI_DATA(PyTypeObject) PyCFunction_Type;
 
-#define PyCFunction_Check(op) Py_IS_TYPE(op, &PyCFunction_Type)
+#define PyCFunction_Check(op) (Py_IS_TYPE(op, &PyCFunction_Type) || (PyType_IsSubtype(Py_TYPE(op), &PyCFunction_Type)))
```

Comment:
BTW, I think we need a `PyCFunction_CheckExact()` macro now if we allow subtypes, to distinguish type checks that guard struct access (Check) from those that optimise for known behaviour (CheckExact).

That probably means that we'll have to look through all usages in CPython to see which is intended. 8-]
This would generally be great to have for Cython, because then I can make Cython's own function type inherit from `PyCFunction_Type` in Py3.9+. Hmm, guess I should volunteer to do the work here…

Also, the usual pattern for `PyXYZ_Check()` functions is to check for a type flag, rather than running a subtype check. Should we add one for PyCFunction, too?

Reply:
Looks like there weren't all that many such type checks. I pushed #20024.

Reply:
> Also, the usual pattern for `PyXYZ_Check()` functions is to check for a type flag, rather than running a subtype check. Should we add one for `PyCFunction`, too?

No. Subclasses of `PyCFunction` are still quite special.
The C-level check should really be used only for very specific optimizations (or the check for `print` in order to give a helpful meessage, I guess). To check for "native callables", look for vectorcall.